### ISSUE TITLE
Session management

### ISSUE THREAD

**Issue by anhchienhoang** [dapphp/securimage#33]
It should have ability to manage the session object in some cases. e.g

$secureImage = new Securimage(array('session_object' => $sessionObject));

if the session_object is not provided then it will use $_SESSION.

By doing this way it's easier to mock that object for testing also.

To ensure the passed session object (or storage object) is valid, you can write an interface for it :)

**Comment by dapphp** [comment]
Upcoming version 4.0 has defined the concept of a "StorageAdapter" to do just this.  I've written adapters for standard PHP sessions, PDO (MySQL, Sqlite, Postgres), Mysqli, Memcached, and Redis.

This new functionality will allow for a custom session adapter to be written as requested in the original issue.

**Comment by peterdd** [comment]
The software (Flyspray) in which I want integrate (using composer.json) securimage uses her own session_name. 
So with 3.6.5 need to insert a config.inc.php with setting the session_name.

So I forked securimage just to be able to hardcode a config.inc.php there.
Any better solution? Should I write my own securimage_play.php and securimage_show.php outside the vendor/ directory?
Because securimage_play.php and securimage_play.php do a new Securimage() without passing any parameter array.

The software is using ADOdb, but I don't think at the moment it is necessary to use db tables and should be ok with just $_SESSION.

**Comment by dapphp** [comment]
Making your own custom securimage_show and securimage_play files is certainly an option.  A lot of framework integrations do this (by having certain routes for captcha image and audio go through the front controller) so the captcha code can have custom settings passed.  My WordPress plugin uses custom files for both (e.g. https://plugins.trac.wordpress.org/browser/securimage-wp/tags/3.6.7/lib/siwp_captcha.php).

I realized lately that it's necessary to have an alternate way to specify a config file if it's not in the securimage directory but not sure the best way.  Maybe something like `ini_set('securimage.config_file', '/path/to/config.inc.php');` but this would have to be set at the vhost or server level, or by middleware/service that runs on all requests.

As the new version is getting closer to beta / release I want to make this available.  If you have any thoughts I'm open to suggestions.

**Comment by peterdd** [comment]
I learned from your Roundcube-Plugin-login_security about 'post-update-cmd' and 
added there:
```
"scripts": {
  "post-update-cmd": [
    "rm -rf vendor/dapphp/securimage/captcha.html .....",
    "echo '<?php return array(\"session_name\"=>\"customsessionname\"); ?>' > vendor/dapphp/securimage/config.inc.php"
  ]
}
```

**Comment by dapphp** [comment]
Good solution but now I realize requiring the config.inc.php file to be in the securimage directory might not always be possible/desirable.  The upcoming version (currently in branch nextgen) allows you to pass a `config_file` key to `Securimage::__construct()` so you can specify a path to the config.

When using some kind of MVC where the captcha image is served through a controller, rather than securimage_show.php and then checked by the application, this approach seems to work well.

Thanks for the suggestion, and good use of the post-update-cmd!

**Comment by dapphp** [comment]
The `config_file` parameter is now available in 3.6.7.
